library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/api/client/TooltipComponentCallback.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2016, 2017, 2018, 2019 FabricMC
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.tooltip.api.client;
+
+import net.fabricmc.api.EnvType;
+import net.fabricmc.api.Environment;
+import net.minecraft.client.gui.tooltip.TooltipComponent;
+import net.minecraft.client.item.TooltipData;
+import net.minecraft.item.Item;
+import org.jetbrains.annotations.Nullable;
+import org.quiltmc.qsl.base.api.event.Event;
+import org.quiltmc.qsl.base.api.event.client.ClientEventAwareListener;
+
+/**
+ * Allows registering a mapping from {@link TooltipData} to {@link TooltipComponent}.
+ * This allows custom tooltips for items: first, override {@link Item#getTooltipData} and return a custom {@link TooltipData}.
+ * Second, register a listener to this event and convert the data to your component implementation if it's an instance of your data class.
+ * <p>
+ * Note that failure to map some data to a component will throw an exception,
+ * so make sure that any data you return in {@link Item#getTooltipData} will be handled by one of the callbacks.
+ */
```

Comment:
This should also have a reference to the ConvertibleTooltipData helper interface